File: security/credential-manager.js
#!/usr/bin/env node
/**
 * Secure Credential Manager for Node.js/Lambda
 * Handles encrypted credential storage and AWS Secrets Manager integration
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const os = require('os');
const { promisify } = require('util');

// AWS SDK (optional, for Lambda deployments)
let AWS;
try {
  AWS = require('aws-sdk');
} catch (e) {
  console.log('AWS SDK not available, using local storage only');
}

class SecureCredentialManager {
  constructor(options = {}) {
    this.vaultPath = options.vaultPath || path.join(os.homedir(), '.my-workspace-vault');
    this.useAWS = options.useAWS && AWS !== undefined;
    this.awsRegion = options.awsRegion || process.env.AWS_REGION || 'us-east-1';

    // Initialize AWS Secrets Manager if available
    if (this.useAWS) {
      this.secretsManager = new AWS.SecretsManager({ region: this.awsRegion });
    }

    // Ensure vault directory exists with proper permissions
    this.initializeVault();

    // Initialize encryption
    this.encryptionKey = this.getOrCreateEncryptionKey();

    // Audit log
    this.auditLog = [];
  }

  /**
   * Initialize vault directory with secure permissions
   */
  initializeVault() {
    if (!fs.existsSync(this.vaultPath)) {
      fs.mkdirSync(this.vaultPath, { recursive: true, mode: 0o700 });
    }

    // Set secure permissions on existing directory
    try {
      fs.chmodSync(this.vaultPath, 0o700);
    } catch (err) {
      console.warn(`Could not set permissions on ${this.vaultPath}:`, err.message);
    }
  }

  /**
   * Get or create encryption key
   */
  getOrCreateEncryptionKey() {
    const keyFile = path.join(this.vaultPath, '.key');

    if (fs.existsSync(keyFile)) {
      const key = fs.readFileSync(keyFile);
      return key;
    }

    // Generate new key
    const key = crypto.randomBytes(32);
    fs.writeFileSync(keyFile, key);
    fs.chmodSync(keyFile, 0o600);

    return key;
  }

  /**
   * Encrypt data
   */
  encrypt(text) {
    const iv = crypto.randomBytes(16);
    const cipher = crypto.createCipheriv('aes-256-cbc', this.encryptionKey, iv);
    let encrypted = cipher.update(text, 'utf8', 'hex');
    encrypted += cipher.final('hex');
    return iv.toString('hex') + ':' + encrypted;
  }

  /**
   * Decrypt data
   */
  decrypt(text) {
    const parts = text.split(':');
    const iv = Buffer.from(parts.shift(), 'hex');
    const encryptedText = Buffer.from(parts.join(':'), 'hex');
    const decipher = crypto.createDecipheriv('aes-256-cbc', this.encryptionKey, iv);
    let decrypted = decipher.update(encryptedText);
    decrypted = Buffer.concat([decrypted, decipher.final()]);
    return decrypted.toString();
  }

  /**
   * Store credential locally
   */
  async storeCredentialLocal(service, key, value, metadata = {}) {
    const credFile = path.join(this.vaultPath, 'credentials.json');

    let credentials = {};
    if (fs.existsSync(credFile)) {
      const encrypted = fs.readFileSync(credFile, 'utf8');
      try {
        const decrypted = this.decrypt(encrypted);
        credentials = JSON.parse(decrypted);
      } catch (err) {
        console.warn('Could not decrypt existing credentials, starting fresh');
        credentials = {};
      }
    }

    // Store credential
    if (!credentials[service]) {
      credentials[service] = {};
    }

    credentials[service][key] = {
      value: value,
      createdAt: new Date().toISOString(),
      rotateBy: new Date(Date.now() + (metadata.rotateDays || 30) * 24 * 60 * 60 * 1000).toISOString(),
      metadata: metadata
    };

    // Save encrypted
    const encrypted = this.encrypt(JSON.stringify(credentials, null, 2));
    fs.writeFileSync(credFile, encrypted);
    fs.chmodSync(credFile, 0o600);

    // Audit log
    this.logAudit('CREDENTIAL_STORED', { service, key });

    return true;
  }

  /**
   * Store credential in AWS Secrets Manager
   */
  async storeCredentialAWS(service, key, value, metadata = {}) {
    const secretName = `my-workspace/${service}`;

    try {
      // Get existing secret or create new one
      let existingSecret = {};
      try {
        const data = await this.secretsManager.getSecretValue({ SecretId: secretName }).promise();
        existingSecret = JSON.parse(data.SecretString);
      } catch (err) {
        if (err.code !== 'ResourceNotFoundException') {
          throw err;
        }
      }

      // Update secret
      existingSecret[key] = value;

      // Store in Secrets Manager
      try {
        await this.secretsManager.updateSecret({
          SecretId: secretName,
          SecretString: JSON.stringify(existingSecret)
        }).promise();
      } catch (err) {
        if (err.code === 'ResourceNotFoundException') {
          // Create new secret
          await this.secretsManager.createSecret({
            Name: secretName,
            SecretString: JSON.stringify(existingSecret),
            Description: `Credentials for ${service}`,
            Tags: [
              { Key: 'Application', Value: 'my-workspace' },
              { Key: 'Service', Value: service },
              { Key: 'ManagedBy', Value: 'credential-manager' }
            ]
          }).promise();
        } else {
          throw err;
        }
      }

      // Audit log
      this.logAudit('CREDENTIAL_STORED_AWS', { service, key });

      return true;
    } catch (err) {
      console.error('Failed to store credential in AWS:', err);
      throw err;
    }
  }

  /**
   * Store credential (auto-selects storage method)
   */
  async storeCredential(service, key, value, metadata = {}) {
    // Always store locally as backup
    await this.storeCredentialLocal(service, key, value, metadata);

    // Also store in AWS if configured
    if (this.useAWS) {
      try {
        await this.storeCredentialAWS(service, key, value, metadata);
      } catch (err) {
        console.warn('Failed to store in AWS, local storage successful');
      }
    }

    return true;
  }

  /**
   * Get credential from local storage
   */
  async getCredentialLocal(service, key) {
    const credFile = path.join(this.vaultPath, 'credentials.json');

    if (!fs.existsSync(credFile)) {
      return null;
    }

    try {
      const encrypted = fs.readFileSync(credFile, 'utf8');
      const decrypted = this.decrypt(encrypted);
      const credentials = JSON.parse(decrypted);

      if (credentials[service] && credentials[service][key]) {
        const entry = credentials[service][key];

        // Check rotation
        const rotateBy = new Date(entry.rotateBy);
        if (rotateBy < new Date()) {
          console.warn(`Credential ${service}:${key} needs rotation (expired ${rotateBy})`);
        }

        // Audit log
        this.logAudit('CREDENTIAL_ACCESSED', { service, key });

        return entry.value;
      }
    } catch (err) {
      console.error('Failed to get credential locally:', err);
    }

    return null;
  }

  /**
   * Get credential from AWS Secrets Manager
   */
  async getCredentialAWS(service, key) {
    const secretName = `my-workspace/${service}`;

    try {
      const data = await this.secretsManager.getSecretValue({ SecretId: secretName }).promise();
      const secrets = JSON.parse(data.SecretString);

      if (secrets[key]) {
        // Audit log
        this.logAudit('CREDENTIAL_ACCESSED_AWS', { service, key });
        return secrets[key];
      }
    } catch (err) {
      if (err.code !== 'ResourceNotFoundException') {
        console.error('Failed to get credential from AWS:', err);
      }
    }

    return null;
  }

  /**
   * Get credential (auto-selects storage method)
   */
  async getCredential(service, key) {
    // Try AWS first if configured
    if (this.useAWS) {
      const awsValue = await this.getCredentialAWS(service, key);
      if (awsValue) {
        return awsValue;
      }
    }

    // Fall back to local storage
    return await this.getCredentialLocal(service, key);
  }

  /**
   * Migrate credentials from .env file
   */
  async migrateFromEnv(envFilePath, service) {
    if (!fs.existsSync(envFilePath)) {
      console.warn(`Environment file not found: ${envFilePath}`);
      return;
    }

    const content = fs.readFileSync(envFilePath, 'utf8');
    const lines = content.split('\n');
    const migrated = [];

    for (const line of lines) {
      const trimmed = line.trim();
      if (trimmed && !trimmed.startsWith('#') && trimmed.includes('=')) {
        const [key, ...valueParts] = trimmed.split('=');
        const value = valueParts.join('=').trim().replace(/^["']|["']$/g, '');

        // Determine rotation period
        let rotateDays = 30;
        if (key.toLowerCase().includes('api') || key.toLowerCase().includes('key')) {
          rotateDays = 90;
        }

        await this.storeCredential(service, key.trim(), value, { rotateDays });
        migrated.push(key.trim());
      }
    }

    if (migrated.length > 0) {
      console.log(`Migrated ${migrated.length} credentials from ${envFilePath}`);

      // Backup original file
      const backupPath = envFilePath + '.backup';
      fs.renameSync(envFilePath, backupPath);
      console.log(`Original file backed up to ${backupPath}`);

      // Log migration
      this.logAudit('ENV_FILE_MIGRATED', { service, file: envFilePath, keys: migrated });
    }

    return migrated;
  }

  /**
   * Setup Lambda environment variables
   */
  async setupLambdaEnvironment() {
    const credentials = await this.getAllCredentials();
    const env = {};

    for (const [service, keys] of Object.entries(credentials)) {
      for (const [key, entry] of Object.entries(keys)) {
        // Convert to uppercase environment variable format
        const envKey = `${service.toUpperCase()}_${key.toUpperCase()}`;
        env[envKey] = entry.value;
      }
    }

    return env;
  }

  /**
   * Get all credentials (for listing)
   */
  async getAllCredentials() {
    const credFile = path.join(this.vaultPath, 'credentials.json');

    if (!fs.existsSync(credFile)) {
      return {};
    }

    try {
      const encrypted = fs.readFileSync(credFile, 'utf8');
      const decrypted = this.decrypt(encrypted);
      return JSON.parse(decrypted);
    } catch (err) {
      console.error('Failed to load credentials:', err);
      return {};
    }
  }

  /**
   * Check which credentials need rotation
   */
  async checkRotation() {
    const credentials = await this.getAllCredentials();
    const needsRotation = [];

    for (const [service, keys] of Object.entries(credentials)) {
      for (const [key, entry] of Object.entries(keys)) {
        const rotateBy = new Date(entry.rotateBy);
        if (rotateBy < new Date()) {
          needsRotation.push({
            service,
            key,
            expiredAt: entry.rotateBy,
            daysOverdue: Math.floor((new Date() - rotateBy) / (1000 * 60 * 60 * 24))
          });
        }
      }
    }

    return needsRotation;
  }

  /**
   * Validate file permissions
   */
  validatePermissions() {
    const results = {};
    const paths = [
      this.vaultPath,
      path.join(this.vaultPath, '.key'),
      path.join(this.vaultPath, 'credentials.json'),
      path.join(this.vaultPath, 'audit.log')
    ];

    for (const p of paths) {
      if (fs.existsSync(p)) {
        const stats = fs.statSync(p);
        const mode = (stats.mode & parseInt('777', 8)).toString(8);

        let isSecure;
        if (stats.isDirectory()) {
          isSecure = mode === '700';
        } else {
          isSecure = mode === '600';
        }

        results[p] = { mode, isSecure };

        if (!isSecure) {
          console.warn(`Insecure permissions on ${p}: ${mode}`);
          // Fix permissions
          if (stats.isDirectory()) {
            fs.chmodSync(p, 0o700);
          } else {
            fs.chmodSync(p, 0o600);
          }
          results[p].fixed = true;
        }
      }
    }

    return results;
  }

  /**
   * Log audit event
   */
  logAudit(action, details) {
    const entry = {
      timestamp: new Date().toISOString(),
      action,
      details,
      user: process.env.USER || 'unknown',
      pid: process.pid
    };

    this.auditLog.push(entry);

    // Also write to file
    const auditFile = path.join(this.vaultPath, 'audit.log');
    fs.appendFileSync(auditFile, JSON.stringify(entry) + '\n');

    // Ensure proper permissions
    if (fs.existsSync(auditFile)) {
      try {
        fs.chmodSync(auditFile, 0o600);
      } catch (err) {
        // Ignore permission errors on some systems
      }
    }
  }

  /**
   * Setup secure MCP configuration
   */
  async setupMCPConfig(services = []) {
    const configDir = path.join(os.homedir(), '.config', 'claude');

    if (!fs.existsSync(configDir)) {
      fs.mkdirSync(configDir, { recursive: true, mode: 0o700 });
    }

    const config = {
      mcpServers: {}
    };

    // Configure each service
    for (const service of services) {
      if (service === 'gmail') {
        const token = await this.getCredential('gmail', 'oauth_token');
        if (token) {
          config.mcpServers.gmail = {
            command: 'npx',
            args: ['-y', '@gongrzhe/server-gmail-autoauth-mcp'],
            env: {
              GMAIL_OAUTH_TOKEN: token
            }
          };
        }
      } else if (service === 'todoist') {
        const apiKey = await this.getCredential('todoist', 'api_key');
        if (apiKey) {
          config.mcpServers.todoist = {
            command: 'npx',
            args: ['-y', '@todoist/mcp-server'],
            env: {
              TODOIST_API_TOKEN: apiKey
            }
          };
        }
      } else if (service === 'ynab') {
        const apiKey = await this.getCredential('ynab', 'api_key');
        if (apiKey) {
          config.mcpServers.ynab = {
            command: 'npx',
            args: ['-y', '@ynab/mcp-server'],
            env: {
              YNAB_API_KEY: apiKey
            }
          };
        }
      }
    }

    // Write config file with secure permissions
    const configFile = path.join(configDir, 'claude_code_config.json');
    fs.writeFileSync(configFile, JSON.stringify(config, null, 2));
    fs.chmodSync(configFile, 0o600);

    console.log(`MCP configuration written to ${configFile}`);
    return configFile;
  }
}

// CLI interface
if (require.main === module) {
  const manager = new SecureCredentialManager({
    useAWS: process.env.USE_AWS === 'true'
  });

  const args = process.argv.slice(2);
  const command = args[0];

  async function main() {
    switch (command) {
      case 'store':
        if (args.length < 4) {
          console.error('Usage: credential-manager.js store <service> <key> <value>');
          process.exit(1);
        }
        await manager.storeCredential(args[1], args[2], args[3]);
        console.log(`Stored credential for ${args[1]}:${args[2]}`);
        break;

      case 'get':
        if (args.length < 3) {
          console.error('Usage: credential-manager.js get <service> <key>');
          process.exit(1);
        }
        const value = await manager.getCredential(args[1], args[2]);
        if (value) {
          console.log(value);
        } else {
          console.error(`Credential not found: ${args[1]}:${args[2]}`);
          process.exit(1);
        }
        break;

      case 'migrate':
        if (args.length < 3) {
          console.error('Usage: credential-manager.js migrate <env-file> <service>');
          process.exit(1);
        }
        await manager.migrateFromEnv(args[1], args[2]);
        break;

      case 'check-rotation':
        const needsRotation = await manager.checkRotation();
        if (needsRotation.length > 0) {
          console.log('Credentials needing rotation:');
          for (const cred of needsRotation) {
            console.log(`  - ${cred.service}:${cred.key} (overdue by ${cred.daysOverdue} days)`);
          }
        } else {
          console.log('All credentials are up to date');
        }
        break;

      case 'validate':
        const results = manager.validatePermissions();
        console.log('Permission validation results:');
        for (const [path, info] of Object.entries(results)) {
          const status = info.isSecure ? '✓ SECURE' : (info.fixed ? '⚠️ FIXED' : '✗ INSECURE');
          console.log(`  ${path}: ${status} (${info.mode})`);
        }
        break;

      case 'setup-mcp':
        if (args.length < 2) {
          console.error('Usage: credential-manager.js setup-mcp <service1,service2,...>');
          process.exit(1);
        }
        const services = args[1].split(',');
        await manager.setupMCPConfig(services);
        break;

      default:
        console.log('Usage: credential-manager.js <command> [options]');
        console.log('Commands:');
        console.log('  store <service> <key> <value>  - Store a credential');
        console.log('  get <service> <key>            - Retrieve a credential');
        console.log('  migrate <env-file> <service>   - Migrate from .env file');
        console.log('  check-rotation                 - Check for credentials needing rotation');
        console.log('  validate                       - Validate file permissions');
        console.log('  setup-mcp <services>          - Setup MCP configuration');
        process.exit(1);
    }
  }

  main().catch(err => {
    console.error('Error:', err);
    process.exit(1);
  });
}

module.exports = SecureCredentialManager;

File: servers/security/credential-manager.py
#!/usr/bin/env python3
"""
Secure Credential Manager for My Workspace
Handles encrypted storage and retrieval of sensitive credentials
"""

import os
import json
import base64
import hashlib
import logging
import stat
from pathlib import Path
from typing import Dict, Optional, Any
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.backends import default_backend
import keyring
from datetime import datetime, timedelta

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

class SecureCredentialManager:
    """
    Manages encrypted credentials with file permission hardening
    and audit logging capabilities
    """

    def __init__(self, vault_path: str = None):
        """Initialize the credential manager"""
        self.vault_path = Path(vault_path or os.path.expanduser("~/.my-workspace-vault"))
        self.vault_path.mkdir(parents=True, exist_ok=True)
        self._set_secure_permissions(self.vault_path)

        self.credentials_file = self.vault_path / "credentials.enc"
        self.audit_log_file = self.vault_path / "audit.log"
        self.key_file = self.vault_path / ".key"

        # Initialize encryption key
        self.cipher = self._initialize_encryption()

        # Initialize audit logger
        self._setup_audit_logger()

    def _set_secure_permissions(self, path: Path):
        """Set secure file permissions (0600 for files, 0700 for directories)"""
        try:
            if path.is_file():
                # Owner read/write only (0600)
                os.chmod(path, stat.S_IRUSR | stat.S_IWUSR)
            elif path.is_dir():
                # Owner read/write/execute only (0700)
                os.chmod(path, stat.S_IRUSR | stat.S_IWUSR | stat.S_IXUSR)
            logger.info(f"Set secure permissions on {path}")
        except Exception as e:
            logger.error(f"Failed to set permissions on {path}: {e}")
            raise

    def _initialize_encryption(self) -> Fernet:
        """Initialize or load encryption key"""
        if self.key_file.exists():
            # Load existing key from keyring
            key = self._load_encryption_key()
        else:
            # Generate new key and store securely
            key = self._generate_encryption_key()

        return Fernet(key)

    def _generate_encryption_key(self) -> bytes:
        """Generate a new encryption key"""
        # Get or create a master password (stored in system keyring)
        master_password = keyring.get_password("my-workspace", "master-key")
        if not master_password:
            # Generate a strong random password
            import secrets
            master_password = secrets.token_urlsafe(32)
            keyring.set_password("my-workspace", "master-key", master_password)
            logger.info("Generated new master encryption key")

        # Derive encryption key from master password
        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=32,
            salt=b'my-workspace-salt-2024',  # Use a fixed salt for consistency
            iterations=100000,
            backend=default_backend()
        )
        key = base64.urlsafe_b64encode(kdf.derive(master_password.encode()))

        # Store key reference (not the actual key)
        self.key_file.write_text("keyring:my-workspace:master-key")
        self._set_secure_permissions(self.key_file)

        return key

    def _load_encryption_key(self) -> bytes:
        """Load encryption key from secure storage"""
        master_password = keyring.get_password("my-workspace", "master-key")
        if not master_password:
            raise ValueError("Master key not found in keyring. Vault may be corrupted.")

        # Derive encryption key from master password
        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=32,
            salt=b'my-workspace-salt-2024',
            iterations=100000,
            backend=default_backend()
        )
        return base64.urlsafe_b64encode(kdf.derive(master_password.encode()))

    def _setup_audit_logger(self):
        """Setup audit logging"""
        audit_handler = logging.FileHandler(self.audit_log_file)
        audit_handler.setLevel(logging.INFO)
        audit_formatter = logging.Formatter(
            '%(asctime)s - AUDIT - %(message)s',
            datefmt='%Y-%m-%d %H:%M:%S'
        )
        audit_handler.setFormatter(audit_formatter)

        self.audit_logger = logging.getLogger('audit')
        self.audit_logger.addHandler(audit_handler)
        self.audit_logger.setLevel(logging.INFO)

        # Secure the audit log
        if self.audit_log_file.exists():
            self._set_secure_permissions(self.audit_log_file)

    def _audit_log(self, action: str, details: Dict[str, Any]):
        """Log an audit event"""
        audit_entry = {
            "timestamp": datetime.utcnow().isoformat(),
            "action": action,
            "details": details,
            "user": os.getenv("USER", "unknown"),
            "pid": os.getpid()
        }
        self.audit_logger.info(json.dumps(audit_entry))

    def store_credential(self, service: str, key: str, value: str,
                        rotate_days: int = 30, metadata: Dict = None):
        """
        Store a credential securely

        Args:
            service: Service name (e.g., 'gmail', 'ynab', 'todoist')
            key: Credential key (e.g., 'api_key', 'oauth_token')
            value: The actual credential value
            rotate_days: Days until rotation reminder
            metadata: Additional metadata to store
        """
        # Load existing credentials
        credentials = self._load_credentials()

        # Prepare credential entry
        entry = {
            "value": value,
            "created_at": datetime.utcnow().isoformat(),
            "rotate_by": (datetime.utcnow() + timedelta(days=rotate_days)).isoformat(),
            "metadata": metadata or {}
        }

        # Store credential
        if service not in credentials:
            credentials[service] = {}

        # Check if overwriting existing
        is_update = key in credentials[service]
        credentials[service][key] = entry

        # Save encrypted credentials
        self._save_credentials(credentials)

        # Audit log
        self._audit_log(
            "CREDENTIAL_STORED" if not is_update else "CREDENTIAL_UPDATED",
            {
                "service": service,
                "key": key,
                "rotate_days": rotate_days,
                "has_metadata": bool(metadata)
            }
        )

        logger.info(f"Stored credential for {service}:{key}")

    def get_credential(self, service: str, key: str) -> Optional[str]:
        """
        Retrieve a credential

        Args:
            service: Service name
            key: Credential key

        Returns:
            The credential value or None if not found
        """
        credentials = self._load_credentials()

        if service in credentials and key in credentials[service]:
            entry = credentials[service][key]

            # Check if rotation is needed
            rotate_by = datetime.fromisoformat(entry["rotate_by"])
            if datetime.utcnow() > rotate_by:
                logger.warning(f"Credential {service}:{key} needs rotation (expired {rotate_by})")

            # Audit log
            self._audit_log("CREDENTIAL_ACCESSED", {
                "service": service,
                "key": key
            })

            return entry["value"]

        return None

    def list_credentials(self) -> Dict[str, list]:
        """List all stored credentials (without values)"""
        credentials = self._load_credentials()
        result = {}

        for service, keys in credentials.items():
            result[service] = []
            for key, entry in keys.items():
                rotate_by = datetime.fromisoformat(entry["rotate_by"])
                needs_rotation = datetime.utcnow() > rotate_by

                result[service].append({
                    "key": key,
                    "created_at": entry["created_at"],
                    "rotate_by": entry["rotate_by"],
                    "needs_rotation": needs_rotation,
                    "metadata": entry.get("metadata", {})
                })

        return result

    def check_rotation_needed(self) -> list:
        """Check which credentials need rotation"""
        credentials = self._load_credentials()
        needs_rotation = []

        for service, keys in credentials.items():
            for key, entry in keys.items():
                rotate_by = datetime.fromisoformat(entry["rotate_by"])
                if datetime.utcnow() > rotate_by:
                    needs_rotation.append({
                        "service": service,
                        "key": key,
                        "expired_at": entry["rotate_by"],
                        "days_overdue": (datetime.utcnow() - rotate_by).days
                    })

        return needs_rotation

    def _load_credentials(self) -> Dict:
        """Load and decrypt credentials"""
        if not self.credentials_file.exists():
            return {}

        try:
            encrypted_data = self.credentials_file.read_bytes()
            decrypted_data = self.cipher.decrypt(encrypted_data)
            return json.loads(decrypted_data.decode())
        except Exception as e:
            logger.error(f"Failed to load credentials: {e}")
            return {}

    def _save_credentials(self, credentials: Dict):
        """Encrypt and save credentials"""
        try:
            json_data = json.dumps(credentials, indent=2)
            encrypted_data = self.cipher.encrypt(json_data.encode())

            # Write with secure permissions
            self.credentials_file.write_bytes(encrypted_data)
            self._set_secure_permissions(self.credentials_file)

        except Exception as e:
            logger.error(f"Failed to save credentials: {e}")
            raise

    def migrate_from_env_file(self, env_file_path: str, service: str):
        """
        Migrate credentials from .env file to secure storage

        Args:
            env_file_path: Path to .env file
            service: Service name for these credentials
        """
        env_path = Path(env_file_path)
        if not env_path.exists():
            logger.warning(f"Environment file not found: {env_file_path}")
            return

        migrated = []
        with open(env_path, 'r') as f:
            for line in f:
                line = line.strip()
                if '=' in line and not line.startswith('#'):
                    key, value = line.split('=', 1)
                    key = key.strip()
                    value = value.strip().strip('"').strip("'")

                    # Determine rotation period based on key type
                    if 'token' in key.lower() or 'oauth' in key.lower():
                        rotate_days = 30
                    elif 'api' in key.lower() or 'key' in key.lower():
                        rotate_days = 90
                    else:
                        rotate_days = 180

                    self.store_credential(service, key, value, rotate_days)
                    migrated.append(key)

        if migrated:
            logger.info(f"Migrated {len(migrated)} credentials from {env_file_path}")
            self._audit_log("ENV_FILE_MIGRATED", {
                "service": service,
                "file": str(env_file_path),
                "keys_migrated": migrated
            })

            # Create backup of original .env file
            backup_path = env_path.with_suffix('.env.backup')
            env_path.rename(backup_path)
            logger.info(f"Original .env file backed up to {backup_path}")

    def export_for_lambda(self, service: str, output_file: str = None):
        """
        Export credentials for AWS Lambda deployment

        Args:
            service: Service to export credentials for
            output_file: Optional output file path
        """
        credentials = self._load_credentials()

        if service not in credentials:
            logger.warning(f"No credentials found for service: {service}")
            return None

        # Prepare export format for Lambda environment variables
        export_data = {}
        for key, entry in credentials[service].items():
            export_data[key] = entry["value"]

        if output_file:
            # Write as JSON for AWS Secrets Manager
            output_path = Path(output_file)
            output_path.write_text(json.dumps(export_data, indent=2))
            self._set_secure_permissions(output_path)
            logger.info(f"Exported credentials to {output_file}")

        return export_data

    def validate_permissions(self) -> Dict[str, bool]:
        """Validate that all credential files have correct permissions"""
        results = {}

        for path in [self.vault_path, self.credentials_file, self.audit_log_file, self.key_file]:
            if path.exists():
                stats = path.stat()
                mode = oct(stats.st_mode)[-3:]

                if path.is_file():
                    # Should be 600 (owner read/write only)
                    is_secure = mode == '600'
                else:
                    # Should be 700 (owner read/write/execute only)
                    is_secure = mode == '700'

                results[str(path)] = is_secure

                if not is_secure:
                    logger.warning(f"Insecure permissions on {path}: {mode}")
                    self._set_secure_permissions(path)

        return results


def main():
    """CLI interface for credential management"""
    import argparse

    parser = argparse.ArgumentParser(description="Secure Credential Manager")
    parser.add_argument("action", choices=["store", "get", "list", "migrate", "check-rotation", "validate"],
                       help="Action to perform")
    parser.add_argument("--service", help="Service name")
    parser.add_argument("--key", help="Credential key")
    parser.add_argument("--value", help="Credential value (for store)")
    parser.add_argument("--env-file", help="Path to .env file (for migrate)")
    parser.add_argument("--rotate-days", type=int, default=30, help="Days until rotation")
    parser.add_argument("--vault-path", help="Custom vault path")

    args = parser.parse_args()

    manager = SecureCredentialManager(vault_path=args.vault_path)

    if args.action == "store":
        if not all([args.service, args.key, args.value]):
            print("Error: --service, --key, and --value required for store")
            return
        manager.store_credential(args.service, args.key, args.value, args.rotate_days)
        print(f"Stored credential for {args.service}:{args.key}")

    elif args.action == "get":
        if not all([args.service, args.key]):
            print("Error: --service and --key required for get")
            return
        value = manager.get_credential(args.service, args.key)
        if value:
            print(value)
        else:
            print(f"Credential not found: {args.service}:{args.key}")

    elif args.action == "list":
        credentials = manager.list_credentials()
        for service, keys in credentials.items():
            print(f"\n{service}:")
            for key_info in keys:
                rotation_status = "⚠️ NEEDS ROTATION" if key_info["needs_rotation"] else "✓"
                print(f"  - {key_info['key']}: expires {key_info['rotate_by']} {rotation_status}")

    elif args.action == "migrate":
        if not all([args.env_file, args.service]):
            print("Error: --env-file and --service required for migrate")
            return
        manager.migrate_from_env_file(args.env_file, args.service)

    elif args.action == "check-rotation":
        needs_rotation = manager.check_rotation_needed()
        if needs_rotation:
            print("Credentials needing rotation:")
            for cred in needs_rotation:
                print(f"  - {cred['service']}:{cred['key']} (overdue by {cred['days_overdue']} days)")
        else:
            print("All credentials are up to date")

    elif args.action == "validate":
        results = manager.validate_permissions()
        print("Permission validation results:")
        for path, is_secure in results.items():
            status = "✓ SECURE" if is_secure else "⚠️ FIXED"
            print(f"  {path}: {status}")


if __name__ == "__main__":
    main()